src/cpu/flexcpu/flexcpu_thread.cc

```diff
@@ -0,0 +1,1673 @@
+void
+FlexCPUThread::recordCycleStats()
+{
+    activeInstructions.sample(inflightInsts.size());
+    if (squashedThisCycle > 0) {
+        squashedPerCycle.sample(squashedThisCycle);
+        squashedThisCycle = 0;
+    }
+    if (fetchedThisCycle > 0) {
+        fetchedInstsPerCycle.sample(fetchedThisCycle);
+        fetchedThisCycle = 0;
+    }
+}
+
+void
+FlexCPUThread::recordInstStats(const shared_ptr<InflightInst>& inst)
+{
+    const InflightInst::TimingRecord& rec = inst->getTimingRecord();
+
+    assert(inst->status() > InflightInst::Status::Invalid);
+
+    if (inst->isSquashed()) {
+        instLifespans.sample(rec.squashTick - rec.creationTick);
+        squashedInstLifespans.sample(rec.squashTick - rec.creationTick);
+        squashedStage[inst->status()]++;
+        return;
+    }
+
+    assert(inst->status() == InflightInst::Status::Committed);
+
+    instLifespans.sample(rec.commitTick - rec.creationTick);
+    committedInstLifespans.sample(rec.commitTick - rec.creationTick);
+    issuedToCommitLatency.sample(rec.commitTick - rec.issueTick);
+    completeToCommitLatency.sample(rec.commitTick - rec.completionTick);
+
+    creationToDecodedLatency.sample(rec.decodeTick - rec.creationTick);
+    decodedToIssuedLatency.sample(rec.issueTick - rec.decodeTick);
+    issuedToExecutingLatency.sample(rec.beginExecuteTick - rec.issueTick);
+    executingToCompleteLatency.sample(rec.completionTick -
+                                      rec.beginExecuteTick);
+
+    if (inst->staticInst()->isMemRef()) {
+        executingToEffAddredLatency.sample(rec.effAddredTick -
+                                           rec.beginExecuteTick);
+        effAddredToMemoryingLatency.sample(rec.beginMemoryTick -
+                                           rec.effAddredTick);
+        memoryingToCompleteLatency.sample(rec.completionTick -
+                                          rec.beginMemoryTick);
+    }
+}
+
+void
+FlexCPUThread::sendToMemory(weak_ptr<InflightInst> inst,
+                            const RequestPtr& req, bool write,
+                            shared_ptr<uint8_t> data,
+                            shared_ptr<SplitRequest> sreq)
+{
+    const shared_ptr<InflightInst> inst_ptr = inst.lock();
+    if (!inst_ptr || inst_ptr->isSquashed()) {
+        // No need to do anything for an instruction that has been squashed.
+        return;
+    }
+
+    sendToMemory(inst_ptr, req, write, data, sreq);
+}
+
+void
+FlexCPUThread::sendToMemory(shared_ptr<InflightInst> inst_ptr,
+                            const RequestPtr& req, bool write,
+                            shared_ptr<uint8_t> data,
+                            shared_ptr<SplitRequest> sreq)
+{
+    inst_ptr->notifyMemorying();
+
+    weak_ptr<InflightInst> weak_inst(inst_ptr);
+
+    if (write) {
+        PacketPtr resp = Packet::createWrite(sreq ? sreq->main : req);
+        if (resp->hasData()) {
+            resp->dataStatic(data.get());
+        }
+        resp->makeResponse();
+
```

Comment:
// Bookkeeping for the thread. Assume that stores complete immediately so we can commit immediately. We assume that completeAcc will be called for all memory instructions.